hibernate-rx-core/src/test/java/org/hibernate/rx/CascadeComplicatedTest.java

```diff
@@ -0,0 +1,442 @@
+package org.hibernate.rx;
+
+import io.vertx.ext.unit.TestContext;
+import org.hibernate.cfg.Configuration;
+import org.hibernate.Hibernate;
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.persistence.*;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletionStage;
+
+/**
+ * This test uses a complicated model that requires Hibernate to delay
+ * inserts until non-nullable transient entity dependencies are resolved.
+ *
+ * All IDs are generated from a sequence.
+ *
+ * JPA cascade types are used (javax.persistence.CascadeType)..
+ *
+ * This test uses the following model:
+ *
+ * <code>
+ *     ------------------------------ N G
+ *     |
+ *     |                                1
+ *     |                                |
+ *     |                                |
+ *     |                                N
+ *     |
+ *     |         E N--------------0,1 * F
+ *     |
+ *     |         1                      N
+ *     |         |                      |
+ *     |         |                      |
+ *     1         N                      |
+ *     *                                |
+ *     B * N---1 D * 1------------------
+ *     *
+ *     N         N
+ *     |         |
+ *     |         |
+ *     1         |
+ *               |
+ *     C * 1-----
+ *</code>
+ *
+ * In the diagram, all associations are bidirectional;
+ * assocations marked with '*' cascade persist, save, merge operations to the
+ * associated entities (e.g., B cascades persist to D, but D does not cascade
+ * persist to B);
+ *
+ * b, c, d, e, f, and g are all transient unsaved that are associated with each other.
+ *
+ * When persisting b with ORM, the entities are added to the ActionQueue in the following order:
+ * c, d (depends on e), f (depends on d, g), e, b, g.
+ *
+ * Entities are inserted in the following order:
+ * c, e, d, b, g, f.
+ */
+public class CascadeComplicatedTest extends BaseRxTest {
+
+	private B b;
+	private C c;
+	private D d;
+	private E e;
+	private F f;
+	private G g;
+
+	@Override
+	protected Configuration constructConfiguration() {
+		Configuration configuration = super.constructConfiguration();
+		configuration.addAnnotatedClass( B.class );
+		configuration.addAnnotatedClass( C.class );
+		configuration.addAnnotatedClass( D.class );
+		configuration.addAnnotatedClass( E.class );
+		configuration.addAnnotatedClass( F.class );
+		configuration.addAnnotatedClass( G.class );
+		return configuration;
+	}
+
+	@Test
+	public void testPersist(TestContext context) {
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.persist(b))
+						.thenCompose(s -> s.flush())
+						.thenCompose(ignore -> check( openSession(), context ))
+		);
+	}
+
+	@Test
+	public void testMergeTransient(TestContext context) {
+		test(
+				context,
+				openSession()
+						.thenCompose(s -> s.merge(b)
+								.thenApply(bMerged -> s)
+						)
+						.thenCompose(s -> s.flush())
```

Comment:
`flush` results in:
`org.hibernate.TransientPropertyValueException: object references an unsaved transient instance - save the transient instance before flushing : org.hibernate.rx.CascadeComplicatedTest$D.e -> org.hibernate.rx.CascadeComplicatedTest$E`